src/Model/Resolver/Product/Options.php

```diff
@@ -43,23 +45,85 @@ class Options extends CoreOptions
     /** @var Uid */
     protected $uidEncoder;
 
+    /**
+     * @var CalculateCustomOptionCatalogRule
+     */
+    protected $calculateCustomOptionCatalogRule;
+
     /**
      * @param PriceCurrencyInterface $priceCurrency
      * @param CatalogData $catalogData
      * @param Uid|null $uidEncoder
+     * @param CalculateCustomOptionCatalogRule|null $calculateCustomOptionCatalogRule
      */
     public function __construct(
         PriceCurrencyInterface $priceCurrency,
         CatalogData $catalogData,
-        Uid $uidEncoder = null
+        Uid $uidEncoder = null,
+        CalculateCustomOptionCatalogRule $calculateCustomOptionCatalogRule
     )
     {
+        $this->calculateCustomOptionCatalogRule = $calculateCustomOptionCatalogRule;
         $this->priceCurrency = $priceCurrency;
         $this->catalogData = $catalogData;
         $this->uidEncoder = $uidEncoder ?: ObjectManager::getInstance()
             ->get(Uid::class);
     }
 
+    /**
+     * @param $price
+     * @param $isPercent
+     * @param $product
+     * @return float
+     */
+    public function getPrice($price, $isPercent, $product)
+    {
+        $catalogPriceValue = $this->calculateCustomOptionCatalogRule->execute(
+            $product,
+            (float)$price,
+            $isPercent
+        );
+
+        if ($catalogPriceValue!==null) {
+            return $catalogPriceValue;
+        }
+
+        return $price;
+    }
+
+    /**
+     * @param array $optonArray
+     * @param $optionValue
+     * @param $product
+     * @param $currentCurrency
+     */
+    public function updateOptionPriceData(array &$optionArray, $optionValue, $product, $currentCurrency) {
+        $optionArray['price_type'] = $optionValue->getPriceType() !== null
+            ? strtoupper($optionValue->getPriceType())
+            : 'DYNAMIC';
+        $optionArray['price'] = $this->getPrice(
+            $optionArray['price'],
+            strtoupper($optionValue->getPriceType()) == 'PERCENT',
```

Comment:
Move please 'PERCENT' into contant or if it is possible then reuse it from 'Magento\Tax\Model\TaxDetails\AppliedTax' which I guess semantically can be correct out here.